### PR TITLE
Middleware refactor

### PR DESCRIPTION
**Pull request by syrusakbary**
# Description
In order to be able to refactor the LLVM experimental compiler, receiving a callback instead of a serialized module metadata, we need to make compilers act on non-mutable input.
This is something that compilers should do, so this PR moves towards the scenario of pure compilation.

# Review

- [ ] Add a short description of the change to the CHANGELOG.md file

### CODE REVIEW

**Comment by Hywan** [comment]
I like the direction of this PR. Is it ready for a review?

**Comment by syrusakbary** [comment]
@Hywan yup, it's ready to review!

bors try

**Comment by syrusakbary** [comment]
bors try-

File: lib/compiler-cranelift/src/config.rs [hunk 1]
```diff
@@ -186,6 +186,11 @@ impl Cranelift {
 
         settings::Flags::new(flags)
     }
+
+    /// Get Middlewares
+    pub fn get_middlewares(&self) -> Vec<Arc<dyn ModuleMiddleware>> {
+        self.middlewares.clone()
+    }
```
Comment (MarkMcCaskey) [inline lib/compiler-cranelift/src/config.rs:193]: Consider using `&[Arc<dyn ModuleMiddleware>]` it won't clone every `Arc` or allocate a new container when this is called. Callers can then get vecs of it with `.to_vec()`.

The primary reason to use `Vec` over `&[]` in a trait method is that you don't want to impose on the implementors that they should actually be storing the data that they're returning. `Vec` lets you just build it and return it, `&[]` requires that the data actually lives on the thing the method is being called on (and that the data is stored in such a way that you can return a `&[]` to it, i.e. linear storage like a Vec or an array).

File: lib/compiler-cranelift/src/compiler.rs [hunk 2]
```diff
@@ -118,15 +121,20 @@ impl Compiler for CraneliftCompiler {
                 // if generate_debug_info {
                 //     context.func.collect_debug_info();
                 // }
+                let mut reader =
+                    MiddlewareBinaryReader::new_with_offset(input.data, input.module_offset);
+                reader.set_middleware_chain(
+                    self.config
+                        .middlewares
+                        .generate_function_middleware_chain(*i),
```
Comment (MarkMcCaskey) [inline lib/compiler-cranelift/src/compiler.rs:129]: I haven't seen the whole PR yet but just looking at this, I'm not quite sure what this does. If I had to guess, it's compiler codegen implemented as a middleware

File: lib/engine-jit/src/artifact.rs [hunk 3]
```diff
@@ -84,10 +84,15 @@ impl JITArtifact {
 
         let compiler = inner_jit.compiler()?;
 
+        let mut module = (*compile_info.module).clone();
+        let middlewares = compiler.get_middlewares();
+        middlewares.apply_on_module_info(&mut module);
```
Comment (nlewycky) [inline lib/engine-jit/src/artifact.rs:89]: This `module` is a clone of the one in `translation`. That means that if the middleware introduces different memory or table styles, our maps `memory_styles` and `table_styles` will be out of date. I suppose the only way to do this today would be for the middleware to add a memory to a module that doesn't have one, or remove memory from one that does. Either way, maybe we should apply the middleware after `let translation = ` above, before we use `translation.module`?

**Review by MarkMcCaskey — APPROVED** [review]
I don't have enough context on middleware itself to have much of a strong opinion about this change, it looks reasonable to me though.

All the Rust stuff looks good, though it might be better to change from `Vec` to `&[]` if it makes sense to.

File: lib/engine-jit/src/artifact.rs (hunk 3 above)
Comment (syrusakbary) [reply lib/engine-jit/src/artifact.rs:89]: I'm just moving the logic, but your comment is completely on point (it was an issue also in previous code).
I'll try to move it up!

Comment (syrusakbary) [reply lib/engine-jit/src/artifact.rs]: Fixed it!

File: lib/compiler-cranelift/src/config.rs (hunk 1 above)
Comment (syrusakbary) [reply lib/compiler-cranelift/src/config.rs]: Fixed!

**Pull request merged**